File: BasicSample/app/src/main/java/com/example/android/persistence/BasicApp.java
/*
 * Copyright 2017, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.android.persistence;

import android.app.Application;
import com.example.android.persistence.db.AppDatabase;

/**
 * Android Application class. Used for accessing singletons.
 */
//自定义 Application 单例模式
public class BasicApp extends Application {

    private AppExecutors mAppExecutors;

    @Override
    public void onCreate() {
        super.onCreate();
        //线程池
        mAppExecutors = new AppExecutors();
    }

    public AppDatabase getDatabase() {
        return AppDatabase.getInstance(this, mAppExecutors);
    }
    //初始化 DataRepository
    public DataRepository getRepository() {
        return DataRepository.getInstance(getDatabase());
    }
}

File: BasicSample-kotlin/app/src/main/java/com/kotlin/basicsample_kotlin/BasicApp.kt
package com.kotlin.basicsample_kotlin

import android.app.Application
import com.kotlin.basicsample_kotlin.db.AppDatabase

class BasicApp : Application() {

    private var appExecutors: AppExecutors? = null

    override fun onCreate() {
        super.onCreate()
        appExecutors = AppExecutors()
    }

    fun database() = AppDatabase.getInstance(this, appExecutors!!)
    fun getRepository() = DataRepository.getInstance(database())
}